File: traductor.py

from typing_extensions import Tuple, List
from deep_translator import GoogleTranslator
import pandas as pd
import langid as lid
import time

def get_file_text(address: str) -> str:
    with open(address, "r", encoding="utf-8") as text:
        csv_text = text.read()
        return csv_text
    

def traducir_texto(texto: str, source_lang: str = "auto", target_lang: str = "en") -> str:
    traductor = GoogleTranslator(source=source_lang, target=target_lang)
    traduccion = traductor.translate(texto)
    return traduccion
    # for i in range(0, len(texto), max_chars):
        # parte = texto[i : i + max_chars]
        # traduccion = traductor.translate(parte)
        # traducciones.append(traduccion)
    # texto_traducido = " ".join(traducciones)
    # return texto_traducido


def is_english(text: str) -> bool:
    lang, num = lid.classify(text)
    return lang == "en"


def is_spanish(text: str) -> bool:
    lang, num = lid.classify(text)
    return lang == "es"


def traducir(
        csv_address: str = "valencia_reviews.csv", 
        nombre_columna: str = "comments", 
        destino: str = "en_reviews.txt", 
        i: int = 0, 
        amount: int = -1
        ) -> Tuple[int, int, List[int], int]:
    
    
    csv = pd.read_csv(csv_address)
    comentarios = []
    serie_comentarios = csv[nombre_columna]
    com = i
    top = i + amount
    count_spanish = 0
    errores = []
    if amount < 0:
        amount = len(serie_comentarios)
    while i < len(serie_comentarios):
        comentario = serie_comentarios[i]
        i += 1
        comentario = str(comentario)
        spanish = is_spanish(comentario)
        if spanish:
            count_spanish += 1
        if is_english(comentario):
            pass
        else:
            comentario = traducir_texto(comentario)
        print(i)
        
        with open(destino, "a", encoding="utf-8") as doc:
            try:
                doc.write(comentario + "¿" + str(spanish) + "\n")
            except:
                errores.append(i)
        if i == top:
            break
    #return "Hecho\n", "errores: ", errores, "comienzo:", com, "fin:", i
    return com, i, errores, count_spanish
a = time.time()
print(traducir())
b = time.time()
print("Tiempo de hacer todos:", b-a)
#print(is_spanish("Hola, buenos días a todos, me llamo Rafael"))

File: analisis_exploratorio/traductor.R
library(readr)
library(textcat)
library(httr)
library(jsonlite)

# Función para leer un archivo de texto
get_file_text <- function(address) {
  read_file(address)
}

# Función para traducir un texto con LibreTranslate (software libre)
traducir_texto <- function(texto, source_lang = "auto", target_lang = "en") {
  url <- "https://libretranslate.com/translate"  # Puedes cambiar a otro servidor si está caído
  body <- list(
    q = texto,
    source = source_lang,
    target = target_lang,
    format = "text"
  )
  
  response <- tryCatch({
    res <- POST(url, body = body, encode = "json", content_type("application/json"))
    json_res <- fromJSON(content(res, "text"))
    return(json_res$translatedText)
  }, error = function(e) {
    print("Errroooooooor")
    return(texto) # Si falla la traducción, devuelve el texto original
  })
  
  return(response)
}

# Función para detectar si el texto está en inglés
is_english <- function(text) {
  lang <- textcat(text)
  return(lang == "english")
}

# Función para detectar si el texto está en español
is_spanish <- function(text) {
  lang <- textcat(text)
  return(lang == "spanish")
}

# Función principal de traducción
traducir <- function(csv_address = "valencia_reviews.csv", 
                     nombre_columna = "comments", 
                     destino = "en_reviews.txt", 
                     i = 1, 
                     amount = -1) {
  
  # Leer CSV
  csv <- read_csv(csv_address, show_col_types = FALSE)
  
  # Obtener comentarios
  comentarios <- csv[[nombre_columna]]
  com <- i
  top <- i + amount
  count_spanish <- 0
  errores <- c()
  
  if (amount < 0) {
    amount <- length(comentarios)
  }
  
  # Traducir comentarios
  while (i <= length(comentarios)) {
    comentario <- as.character(comentarios[i])
    i <- i + 1
    
    spanish <- is_spanish(comentario)
    if (spanish) {
      count_spanish <- count_spanish + 1
    }
    
    if (!is_english(comentario)) {
      comentario <- traducir_texto(comentario)
    }
    
    print(i)
    
    # Guardar en archivo
    tryCatch({
      write(paste0(comentario, "¿", spanish, "¿", i, "\n"), 
            file = destino, append = TRUE)
    }, error = function(e) {
      errores <- c(errores, i)
    })
    
    if (i == top) {
      break
    }
  }
  
  return(list(comienzo = com, fin = i, errores = errores, total_spanish = count_spanish))
}

# Medir tiempo de ejecución
start_time <- Sys.time()
print(traducir())
end_time <- Sys.time()
print(paste("Tiempo total:", end_time - start_time))
